zbus/src/abstractions/process.rs

```diff
@@ -94,3 +96,21 @@ where
 {
     Command::new(program).args(args).output().await
 }
+
+impl From<&Unixexec<'_>> for Command {
+    fn from(unixexec: &Unixexec<'_>) -> Self {
+        let mut command = Command::new(unixexec.path());
+
+        if let Some(arg0) = unixexec.find_arg_by_key(0) {
+            command.arg0(arg0.as_ref());
+        }
+
+        let mut arg_index = 1;
+        while let Some(arg) = unixexec.find_arg_by_key(arg_index) {
+            command.arg(arg.as_ref());
+            arg_index += 1;
+        }
```

Comment:
Agree and I would like to get rid of it if possible. However, I haven't found a more idiomatic way to rewrite it (by iterating over `args`) that:
- gets rid of duplicates
- stops at the first "gap" in the index array

While getting rid of duplicates could be done by converting `args` into a `HashMap` first, I'm not sure what's a good way of detecting the first "gap".

TLDR: I haven't found a cleaner way to rewrite it (that's still spec-compliant) but I'm open to suggestions for improving it.

Reply:
> However, I haven't found a more idiomatic way to rewrite it (by iterating over `args`) that:
> 
> * gets rid of duplicates

Hmm.. I think that's a more general issue since other address types also use key and value. If we want to solve this, we need a more general solution here. However, as I wrote on #989, we should just ignore duplicates and take the last value.

> * stops at the first "gap" in the index array

Ah right, the spec explicitly says how to handle that. Still, I think the code will be nicer if we do it through iterating over all args:

```rust
        let mut last_index = 0;
        for (i, arg) in unixexec.args() {
            if i != last_index + 1 {
                break;
            }

            command.arg(arg.as_ref());
            last_index = i;
        }
```

Reply:
The above code assumes that we'll change arg0 to be treated separately (separate field+getter) as I suggested [here](https://github.com/dbus2/zbus/pull/989#discussion_r1784468781).

Reply:
> ```rust
> let mut last_index = 0;
> for (i, arg) in unixexec.args() {
>     if i != last_index + 1 {
>         break;
>     }
> 
>     command.arg(arg.as_ref());
>     last_index = i;
> }
> ```

The downsides of this code is that a) it assumes the `args()` are sorted (which is somewhat reasonable) and b) it doesn't handle duplicates. Therefore, I've adjusted it a bit in the latest revision so that it handles duplicates.

But (to me at least) the revised code is less readable while still intrusive to the details of a `Unixexec` and C-like, to the point where I"m not sure if it's actually an improvement over the previous implementation.

This is all a bit subjective and the end result should be the same, so it's fine by me either way. I'm still open to further refactor and improve it though, if you have other suggestions.

Reply:
> The downsides of this code is that a) it assumes that `args()` are sorted (which is somewhat reasonable)

I think they should be sorted, regardless of this code. There is `Vec::sort_by` to ensure that. The performance penalty will be 0 if the address string had the args sorted already, which they really should be. If people don't put args in order in the address, they really deserve any small performance penalty from the sorting. :)

But this should happen in the construction/parsing of `Unixexec`. I see that you've modified the `Command` abstraction to be more specific to `Unixexec` needs. I don't think that's necessary. Except for `for_unixexec` method, `Command` should provide a generic API: `arg` builder method to push each arg into it and `for_unixexec` should make use of that.

> and b) it doesn't handle duplicates. Therefore, I've adjusted it a bit in the latest revision so that it handles duplicates.

We already talked about duplicates above. As I wrote that needs a general solution as it's not specific to unixexec. However, if you really want to solve it only for unixexec, you can filter the duplicates in `Unixexec` construction (just like sorting). There is also `Vec::dedup_by`.

Thinking about it, you could as well take care of gaps in the same place. Doing all the processing upfront, will also ensure we only do all this once and any further uses of a parsed `Unixexec` can just assume everything is in order.

> But (to me at least) the revised code is less readable while still intrusive to the details of a `Unixexec` and C-like,

I can understand completely that you find it less readable but it's not intrusive since it's not asking for specific indexes but rather just getting a slice of all args and then iterating over them. Also, C doesn't have for loops that iterate over a collection or tuples, so not sure what's C-like about it. :)

You're right that it will be assuming there are no duplicates and args being sorted etc. However, those assumptions could be part of the API-contract of `Unixexec::args` (could even be documented).

> This is all a bit subjective and the end result should be the same, so it's fine by me either way. I'm still open to further refactor and improve it though, if you have other suggestions.

Sure and you can always disagree with me. However, in the future please feel free to be blunt about it. :)

Reply:
> However, in the future please feel free to be blunt about it. :)

:+1: While I may disagree with small bits and pieces, I still appreciate the feedback so please keep it coming :)

Pushed another revision which moves the deduplication and the removal of the gaps during the `Unixexec` construction.

This has the nice side-effect that it simplified the `Command::for_unixexec` implementation which is pretty clean and readable now.